File: scripts/inject-og-social.py
#!/usr/bin/env python3
"""Injecte Open Graph sur pages articles. Usage: python scripts/inject-og-social.py"""
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BASE = "https://socrate83.github.io/premieres-nations-quebec"
DEFAULT_IMG = f"{BASE}/images/articles/78-friction-feu.png"
LANDING = {"Carnets.html", "Articles.html", "Home.html"}


def meta_content(html: str, name: str) -> str:
    for pat in (
        rf'<meta[^>]+name=["\']{name}["\'][^>]+content=["\']([^"\']*)["\']',
        rf'<meta[^>]+content=["\']([^"\']*)["\'][^>]+name=["\']{name}["\']',
    ):
        m = re.search(pat, html, re.I)
        if m:
            return m.group(1)
    return ""


def first_image(html: str) -> str:
    m = re.search(r'src=["\']images/articles/([^"\']+)["\']', html, re.I)
    return f"images/articles/{m.group(1)}" if m else ""


def strip_og(html: str) -> str:
    html = re.sub(r'\s*<meta property="og:[^"]+"[^>]*>\s*', "", html, flags=re.I)
    html = re.sub(r'\s*<meta name="twitter:(?!site)[^"]+"[^>]*>\s*', "", html, flags=re.I)
    return html


def esc(s: str) -> str:
    return s.replace('"', "&quot;")


def main():
    n = 0
    for fp in sorted(ROOT.glob("*.html")):
        html = fp.read_text(encoding="utf-8")
        if "pn-article-root" not in html and "article-ecoute" not in html and fp.name not in LANDING:
            continue
        title_m = re.search(r"<title>([^<]*)</title>", html, re.I)
        title = title_m.group(1).strip() if title_m else fp.name
        desc = meta_content(html, "description") or title
        img = first_image(html)
        img_abs = f"{BASE}/{img}" if img else DEFAULT_IMG
        url = f"{BASE}/{fp.name}"

        html = strip_og(html)
        block = f"""
  <meta property="og:title" content="{esc(title)}">
  <meta property="og:description" content="{esc(desc)}">
  <meta property="og:url" content="{url}">
  <meta property="og:type" content="article">
  <meta property="og:image" content="{img_abs}">
  <meta property="og:locale" content="fr_FR">
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:title" content="{esc(title)}">
  <meta name="twitter:description" content="{esc(desc)}">
  <meta name="twitter:image" content="{img_abs}">"""

        if "</head>" in html:
            html = html.replace("</head>", block + "\n</head>", 1)
            fp.write_text(html, encoding="utf-8")
            n += 1
            print(f"og: {fp.name}")
    print(f"Terminé — {n} pages.")


if __name__ == "__main__":
    main()

File: scripts/inject-og-social.mjs
/**
 * Injecte balises Open Graph / Twitter sur les pages articles (aperçu Facebook riche).
 * Usage: node scripts/inject-og-social.mjs
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const root = path.join(path.dirname(fileURLToPath(import.meta.url)), '..');
const BASE = 'https://socrate83.github.io/premieres-nations-quebec';

function metaContent(html, name) {
  const re = new RegExp(`<meta[^>]+name=["']${name}["'][^>]+content=["']([^"']*)["']`, 'i');
  const m = html.match(re);
  if (m) return m[1];
  const re2 = new RegExp(`<meta[^>]+content=["']([^"']*)["'][^>]+name=["']${name}["']`, 'i');
  const m2 = html.match(re2);
  return m2 ? m2[1] : '';
}

function firstArticleImage(html) {
  const m = html.match(/src=["']images\/articles\/([^"']+)["']/i);
  return m ? `images/articles/${m[1]}` : '';
}

function stripOg(html) {
  return html.replace(/\s*<meta property="og:[^"]+"[^>]*>\s*/gi, '')
    .replace(/\s*<meta name="twitter:(?!site)[^"]+"[^>]*>\s*/gi, '');
}

let n = 0;
for (const name of fs.readdirSync(root)) {
  if (!name.endsWith('.html')) continue;
  const fp = path.join(root, name);
  let html = fs.readFileSync(fp, 'utf8');
  const isArticle = html.includes('pn-article-root') || html.includes('article-ecoute');
  const isLanding = ['Carnets.html', 'Articles.html', 'Home.html'].includes(name);
  if (!isArticle && !isLanding) continue;

  const title = (html.match(/<title>([^<]*)<\/title>/i) || ['', name])[1].trim();
  const desc = metaContent(html, 'description') || title;
  const img = firstArticleImage(html);
  const imgAbs = img ? `${BASE}/${img}` : `${BASE}/images/articles/78-friction-feu.png`;
  const url = `${BASE}/${name}`;

  html = stripOg(html);
  const block = `
  <meta property="og:title" content="${title.replace(/"/g, '&quot;')}">
  <meta property="og:description" content="${desc.replace(/"/g, '&quot;')}">
  <meta property="og:url" content="${url}">
  <meta property="og:type" content="article">
  <meta property="og:image" content="${imgAbs}">
  <meta property="og:locale" content="fr_FR">
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:title" content="${title.replace(/"/g, '&quot;')}">
  <meta name="twitter:description" content="${desc.replace(/"/g, '&quot;')}">
  <meta name="twitter:image" content="${imgAbs}">`;

  if (html.includes('</head>')) {
    html = html.replace('</head>', `${block}\n</head>`);
    fs.writeFileSync(fp, html, 'utf8');
    n += 1;
    console.log('og:', name);
  }
}
console.log('Terminé —', n, 'pages.');
